### PR TITLE
Add hook for `Timeout.timeout`.

### PR DESCRIPTION
**Pull request by ioquatix**
Also, I'd like to consider renaming it to `Timeout.raise` or `Timeout.after` because I'm not sure the current name makes sense. We can leave existing name as alias.

### CODE REVIEW

File: lib/timeout.rb [hunk 1]
```diff
@@ -76,9 +76,15 @@ def exception(*)
   # Note that this is both a method of module Timeout, so you can <tt>include
   # Timeout</tt> into your classes so they have a #timeout method, as well as
   # a module method, so you can call it directly as Timeout.timeout().
-  def timeout(sec, klass = nil, message = nil)   #:yield: +sec+
+  def timeout(sec, klass = nil, message = nil, &block)   #:yield: +sec+
     return yield(sec) if sec == nil or sec.zero?
+
     message ||= "execution expired".freeze
+
+    if scheduler = Fiber.scheduler and scheduler.respond_to?(:timeout_raise)
```
Comment (nobu) [inline lib/timeout.rb:84]:
```suggestion
    if (scheduler = Fiber.scheduler)&.respond_to?(:timeout_raise)
```

**Comment by nobu** [comment]
`rb_fiber_scheduler_timeout_raise` seems unnecessary.

File: scheduler.c [hunk 2]
```diff
@@ -108,6 +110,12 @@ rb_fiber_scheduler_make_timeout(struct timeval *timeout)
     return Qnil;
 }
 
+VALUE
+rb_fiber_scheduler_timeout_raise(VALUE scheduler, VALUE timeout)
+{
+    return rb_check_funcall(scheduler, id_timeout_raise, 1, timeout);
```
Comment (eregon) [inline scheduler.c]: This doesn't pass the block and doesn't match the arguments of `timeout_raise` above.
Probably best to remove until it's tested.
Also, I see no potential usage for it as a C API.

Comment (nobu) [inline scheduler.c:116]:
```suggestion
    return rb_check_funcall(scheduler, id_timeout_raise, 1, &timeout);
```

**Comment by eregon** [comment]
> Also, I'd like to consider renaming it to `Timeout.raise` or `Timeout.after` because I'm not sure the current name makes sense. We can leave existing name as alias.

Not in this PR please, that needs discussion.
I'm pretty sure everyone is used to to `Timeout.timeout`, so renaming will only hurt or add confusion and I advise against it.

File: scheduler.c (hunk 2 above)
Comment (eregon) [reply scheduler.c]: Looks good to me except this.

Comment (ioquatix) [reply scheduler.c]: Thanks, I didn't realise it had diverged so much, I'll fix it.

File: lib/timeout.rb [hunk 3]
```diff
@@ -76,9 +76,15 @@ def exception(*)
   # Note that this is both a method of module Timeout, so you can <tt>include
   # Timeout</tt> into your classes so they have a #timeout method, as well as
   # a module method, so you can call it directly as Timeout.timeout().
```
Comment (rafaelfranca) [inline lib/timeout.rb:78]: Should the documentation be updated to explain this new behavior?

Comment (ioquatix) [reply lib/timeout.rb:78]: Yes good idea.

**Comment by ioquatix** [comment]
I am going to revisit the public C interface before the next release.

**Pull request merged**